app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+
+            // Update the selected hike if necessary
+            updateSelectedHike()
+          }
+
+          onFailure()
+          return@withContext
+        }
+
+        // Update the hike in the list
+        var success = false
+        _hikesMutex.withLock {
+          val hikeFlow = _hikeFlowsMap[hikeId] ?: return@withLock
+          val updatedHike =
+              hikeFlow.value.copy(
+                  distance = DeferredData.Obtained(distance),
+                  elevationGain = DeferredData.Obtained(elevationGain),
+                  estimatedTime = DeferredData.Obtained(estimatedTime),
+                  difficulty = DeferredData.Obtained(difficulty))
+          hikeFlow.value = updatedHike
+
+          // Update the selected hike if necessary
+          updateSelectedHike()
+
+          success = true
+        }
+
+        if (success) {
+          onSuccess()
+        } else {
+          onFailure()
+        }
+      }
+
+  /**
+   * Helper function for [computeDetailsForAsync].
+   *
+   * Checks that the required data ([Hike.waypoints] and [Hike.elevation]) are available, and that a
+   * request is not already ongoing. Returns the data if it is available, and indicates whether a
+   * request should be performed.
+   *
+   * If a request should be performed, this function will also mark the data as requested.
+   *
+   * Note: this function does not set its context to [dispatcher], it is the responsibility of the
+   * caller to do so.
+   *
+   * @param hikeId The ID of the hike to check the data for.
+   * @param onSuccess To be called if the data was already obtained or requested. Won't be called if
+   *   a request should be performed.
+   * @param onFailure To be called if no hike corresponding to the provided ID was found.
+   * @return A triple containing:
+   * - A `Boolean` indicating whether the request should be performed. A value of true means the
+   *   request should be performed, false means the data is already available or an error occurred.
+   * - The `List` of way points of the hike, if available. If not available, the list will be empty
+   *   and the boolean will be false.
+   * - The `List` of elevation data of the hike, if available. If not available, the list will be
+   *   empty and the boolean will be false.
+   */
+  private suspend fun checkDetailsCanBeRetrievedForAsync(
+      hikeId: String,
+      onSuccess: () -> Unit,
+      onFailure: () -> Unit
+  ): Triple<Boolean, List<LatLong>, List<Double>> {
+    var success = false
+    var alreadyComputed = false
+    var alreadyRequested = false
+    var nullableWayPoints: List<LatLong>? = null
+    var nullableElevation: List<Double>? = null
+    _hikesMutex.withLock {
+      val hikeFlow = _hikeFlowsMap[hikeId] ?: return@withLock
+      val hike = hikeFlow.value
+
+      // Check that the details data hasn't been computed already
+      alreadyComputed = areDetailsComputedFor(hike)
+      if (alreadyComputed) {
+        success = true
```

Comment:
I'm assuming your question is "why do we stop before loading waypoints and elevation when the data is already computed".

This is because waypoints and elevation are needed to compute the details. But if the computation is not needed (data are already requested or obtained), no need for the waypoints and elevation, they won't be used anyway.

Reply:
Nope, my question's idea is more "We have the details from the hike, why don't we return them in the Triple instead of emptyList?"

Reply:
Well actually we do not have them yet, we would have to extract them, which includes going to the end of the function. What I do here is an early return, because since the details were already computed, we do not need to return the data to compute them.

I could return them, it's just not necessary and would require more operations.

Reply:
Mmmmh, I see. Actually, it makes sense now that the function is a "PreRequestOperation" and no longer a check function. Thanks!